proposals/0011-inline-spirv.md

```diff
@@ -0,0 +1,347 @@
+
+The current inline spir-v include the `vk::ext_decorate` attribute. This
+generally works well as an attribute. A header file could handle the attribute
+in the same way that it handles the execution mode attribute.
+
+### Storage classes
+
+The existing inline spir-v allows the developer to set the storage class for a
+variable. Conceptually this is similar to setting an address space. I would not
+want to add anything new, but I suggest we deprecate this when address spaces
+are added to HLSL more generally.
+
+### Builtin input
+
+The existing inline spir-v has limited support for adding a builtin. There is a
+sample that shows how to
+[add a builtin input to a pixel shader](https://github.com/microsoft/DirectXShaderCompiler/blob/128b6fd16b449df696a5c9f9405982903a4f88c4/tools/clang/test/CodeGenSPIRV/spv.intrinsicDecorate.hlsl).
+This work for pixel shaders because the parameters to a pixel shader can have
+arbitrary semantics. However, when we quickly ran into problems when
+[trying to declare a spir-v specific builtin in a compute shader](https://github.com/microsoft/DirectXShaderCompiler/issues/4217).
+
+The reason that a builtin input requires more than just a decoration on the
+variable. It must be in the input storage class, and get added to the
+`OpEntryPoint` instruction. With the existing inline spir-v we can only get the
+first two.
+
+Since a lot (TODO: Add exact numbers) extensions add new builtin inputs, it
+would be good to have a clean way to do this. I'm proposing a new function
+attribute `vk::ext_builtin_input` that, like `vk::ext_instruction`, will imply a
+definition for the function. When the user calls the function, it will return
+the value of that builtin input.
+
+For example, we could declare the `gl_NumWorkGroups` builtin in a header file
+like this:
+
+```
+[[vk::ext_builtin_input(/* NumWorkgroups */ 24)]]
+uint32 gl_NumWorkGroups();
+```
+
+Then the compiler will be able to add a variable to in the input storage class,
+with the builtin decoration, with a type that is the same as the return type of
+the function, and add it to the OpEntryPoint for the entry points from which it
+is reachable.
+
+The developer can use the builtin input by simply calling the function.
+
+### Builtin output
+
+The existing inline spir-v has limited support for adding a builtin. This would
+have most of the same problems as declaring a builtin input.
+
+Nine of the 111 extensions add new builtin outputs. It would be good to have a
+clean way to do this. I'm proposing a new function attribute
+`vk::ext_builtin_ouput(val)` that, like `vk::ext_instruction`, will imply a
+definition for the function. When the user calls the function, it will set the
+value of the output to `val`.
+
+For example, we could declare the `gl_FragStencilRefARB` builtin in a header
+file like this:
+
+```
+[[vk::ext_extension("SPV_EXT_shader_stencil_export")]]
+[[vk::ext_builtin_input(/* FragStencilRefEXT */ 5014)]]
+void gl_FragStencilRefARB(int);
```

Comment:
Why is that needed for built-ins to be used like functions?
Would it be possible to declare the builtin-type like the other types (if needed),
and have a global variable with the buildin annotation?
Or is declaring such global an issue for implementation?

Reply:
That is a good question. That idea could work. The syntax would be:

```
[[vk::ext_extension("SPV_EXT_shader_stencil_export")]]
[[vk::ext_builtin_output(/* FragStencilRefEXT */ 5014)]]
int gl_FragStencilRefARB;
```

Then `gl_FragStencilRefARB` could be used like any other global variable. I think that could work. I'll wait for more feedback before I update the proposal.

Reply:
I will open a follow up issue to consider this. I'll look at implementation issue to determine which is easiest. I'm conderned that there is a lot of code that handled variables based on their representation in the AST, and I'll have to special case it to look for the attribute.